R/make-traceability-matrix.R

```diff
@@ -332,38 +354,58 @@ get_exports <- function(pkg_source_path){
 #'
 #' @keywords internal
 filter_symbol_functions <- function(funcs){
-  ignore_functions <- c("\\%>\\%", "\\$", "\\[\\[", "\\[", "\\+")
-  pattern <- paste0("^(", paste(ignore_functions, collapse = "|"), ")$")
+  ignore_patterns <- c("\\%>\\%", "\\$", "\\[\\[", "\\[", "\\+", "\\%", "<-")
+  pattern <- paste0("(", paste(ignore_patterns, collapse = "|"), ")")
   funcs_return <- grep(pattern, funcs, value = TRUE, invert = TRUE)
   return(funcs_return)
 }
 
-#' list all functions defined in the package code
+#' list all top-level objects defined in the package code
 #'
-#' @inheritParams make_traceability_matrix
+#' This is primarily for getting all _functions_, but it also returns top-level
+#' declarations, regardless of type. This is intentional, because we also want
+#' to capture any global variables or anything else that could be potentially
+#' exported by the package.
+#'
+#' @inheritParams map_functions_to_scripts
 #'
 #' @return A data.frame with the columns `func` and `code_file` with a row for
-#'   every function defined in the package.
+#'   every top-level object defined in the package.
 #'
 #' @keywords internal
-get_all_functions <- function(pkg_source_path){
-
-  # Get all defined functions (following syntax: func <- function(arg), or setGeneric("func"))
-  r_files <- list.files(file.path(pkg_source_path, "R"), full.names = TRUE)
+get_toplevel_assignments <- function(pkg_source_path){
+  r_files <- list.files(
+    file.path(pkg_source_path, "R"),
+    full.names = TRUE, pattern = "\\.[Rr]$", recursive = TRUE
+  )
```

Comment:
Good change. Was going to add this next for the `pbapply` package. Confirmed it worked well for nested `R/` directories

Reply:
> Was going to add this next for the pbapply package.

Thanks for pointing to a package where this is relevant!  I was mentioning to @seth127 the other day that I didn't think code in a subdirectory would make it to the package namespace (and that was based on just creating an arbitrary subdirectory and testing).

pbapply has "unix" and "windows" subdirectories, and it looks like R has special handling for OS subdirectories:

https://github.com/wch/r-source/blob/ef3455e0c5d385d6484f6572ab124de97290f853/src/library/tools/R/utils.R#L159-L185

Reply:
I've pushed an update to the commit message for this commit to avoid spreading misinformation.  No code changes.

<details>
<summary>range-diff</summary>

```
 1:  22e82d2 =  1:  22e82d2 add minor regression test for change made in #33
 2:  f81479a =  2:  f81479a matrix tests: tweak test names
 3:  359f846 =  3:  359f846 map_functions_to_scripts: document missing parameters
 4:  92c9e01 =  4:  92c9e01 get_testing_dir: delete commented code
 5:  400b4a6 =  5:  400b4a6 matrix: restrict 'no testthat directory' to verbose mode
 6:  06a2111 =  6:  06a2111 test 'no R directory' case
 7:  2daf0c3 =  7:  2daf0c3 add support for no `R/` folder - error out
 8:  81e8d00 =  8:  81e8d00 add support for no exports found
 9:  f0a73c9 =  9:  f0a73c9 filter exports based on symbol patterns instead of specific functions
10:  15a8001 = 10:  15a8001 matrix: filter patterns with '%' and '<-' from exports
11:  287c6a5 = 11:  287c6a5 refactor tests to better isolate "no docs" from syntax errors
12:  da136df = 12:  da136df matrix: filter to R files when listing R/
13:  ac0e705 = 13:  ac0e705 matrix: use parse rather than regex to collect top-level functions
14:  79feaf6 = 14:  79feaf6 get_toplevel_assignments: tweak approach constructing code_file name
15:  3184dc4 ! 15:  a97812a get_toplevel_assignments: recursively list files
    @@ Metadata
      ## Commit message ##
         get_toplevel_assignments: recursively list files

    -    [km]: I don't think there is any issue with this change, but, from my
    -          testing, anything defined in an R/ subdirectory will make it
    -          through to the source tarball but will _not_ make it into the
    -          package namespace.
    +    In general code in subdirectories won't make it through to the
    +    namespace of an installed package, but R has special handling for OS
    +    subdirectories (see pbapply for an example of a package that uses
    +    this).

      ## R/make-traceability-matrix.R ##
     @@ R/make-traceability-matrix.R: filter_symbol_functions <- function(funcs){
16:  23f4eb7 = 16:  0507603 map_functions_to_scripts: tweak 'exports were not found' message
17:  c0c9388 = 17:  09c1988 matrix tests: check multi-line function declaration
18:  6bf0156 = 18:  a4a805c matrix tests: add second arg to setGeneric calls
19:  fd9a70c = 19:  0457d9e matrix tests: add non-function top-level assignment case
20:  9b9c2f3 = 20:  7024c91 matrix tests: check exported_function and codefile
21:  4b5deda = 21:  c0e412b add support for exportpatterns
```

</details>

Reply:
@kyleam 
>Thanks for pointing to a package where this is relevant! ...pbapply has "unix" and "windows" subdirectories, and it looks like R has special handling for OS subdirectories

No prob! Yeah this was the only scenario where i've seen this, and was wondering if any directory was acceptable here. Do you think it's necessary to add a test for this scenario? I can add one on top of this PR if that's easier (and we want to)

Reply:
> Do you think it's necessary to add a test for this scenario?

Yeah, I think following up with that at some point is a good idea.  That would also probably force us to think more about the "need to check if there are funcs are defined in multiple places?" todo since the main use case for having the "unix" and "windows" split is for having different implementations of the same 

https://github.com/metrumresearchgroup/mpn.scorecard/blob/c0e412bfe1fd62fef1c467b87267a8d00d77a725/R/make-traceability-matrix.R#L411-L411